src/components/data/_tests/data-utils.test.js

```diff
@@ -428,3 +429,304 @@ describe('prepareAttributeForUpload', () => {
     })
   })
 })
+
+describe('AttributeTypeInput', () => {
+  it('renders radio buttons for available types', () => {
+    const { getAllByRole } = render(h(AttributeTypeInput, {
+      value: { type: 'string' },
+      onChange: jest.fn()
+    }))
+
+    const radioButtons = getAllByRole('radio')
+    const labels = Array.from(radioButtons).map(el => el.labels[0].textContent)
+
+    expect(radioButtons.length).toBe(4)
+    expect(labels).toEqual(['String', 'Reference', 'Number', 'Boolean'])
+  })
+
+  it('renders a radio button for JSON type if requested', () => {
+    const { queryByLabelText } = render(h(AttributeTypeInput, {
+      value: { type: 'string' },
+      onChange: jest.fn(),
+      showJsonTypeOption: true
+    }))
+
+    const jsonRadioButton = queryByLabelText('JSON')
+    expect(jsonRadioButton).not.toBeNull()
+  })
+
+  it('calls onChange callback when a type is selected', () => {
+    const onChange = jest.fn()
+    const { getByLabelText } = render(h(AttributeTypeInput, {
+      value: { type: 'string' },
+      onChange
+    }))
+
+    fireEvent.click(getByLabelText('Number'))
+
+    expect(onChange).toHaveBeenCalledWith({ type: 'number' })
```

Comment:
In my opinion, this is a decent alternative to TypeScript for improving the visibility of a contract. I hope to find opportunities to steal this technique.

Reply:
I think tests like this are good for ensuring that a component continues to adhere to a contract. But I'll claim that if/when a contract changes, TypeScript makes it much easier to check that all consumers have been updated to the new contract.